src/crimson/osd/ops_executer.h

```diff
@@ -504,6 +504,7 @@ OpsExecuter::flush_changes_n_do_ops_effects(
     ceph_assert(want_mutate);
   }
 
+  apply_stats();
```

Comment:
Can you please explain why moving OpsExecutor::obc is relevant to apply_stats?

Reply:
`apply_stats` also tries to access `OpsExecutor::obc`, so it has to called before we "move" `OpsExecutor::obc`. I'll add comments to the code.

Reply:
Not sure where it tries to access it, I might be missing something here. To me it looks we don't actually use obc.

Reply:
Yeah, I think I was wrong about this, will re-examine the case.

Reply:
@Matan-B Hi, matan, it turned out that the case was caused by other conditions and only exist in our test version. I've just removed that commit, please take a look again, thanks:-)

Reply:
@Matan-B Sorry, I was wrong again, apply_stats does access obc, please see the following backtrace:
```
crimson-osd: /home/xuxuehan/src/ceph/clang-build/boost/include/boost/smart_ptr/intrusive_ptr.hpp:201: T *boost::intrusive_ptr<crimson::osd::ObjectContext>::operator->() const [T = crimson::osd::ObjectContext]: Assertion `px != 0' failed.
Aborting on shard 0.
Backtrace:
Reactor stalled for 3359 ms on shard 0, in scheduling group main. Backtrace: 0x78a43d7 0x78a422c 0x783c1b6 0x783acd0 0x783abb0 0x783ae4b 0x783c586 0x4251f 0x174627 0x5b012b0 0x5b017fd 0x5b028b9 0x5b011a6 0x5b00f91 0x5b00be2 0x5b00a4b 0x5b00978 0x5b000c4 0x5affd82 0x5affcc5 0x4251f 0x969fb 0x42475 0x287f2 0x2871a 0x39e95 0x4dc4091 0x4e3bc7b 0x4dc87a4 0x4fdebfc 0x4dba57e 0x4e1cff7 0x4e1cd94 0x7845d61 0x784731a 0x7848352 0x7847734 0x77aa304 0x77a9746 0x4b74c4e 0x29d8f 0x29e3f 0x4ad2fc4
Reactor stalled for 6383 ms on shard 0, in scheduling group main. Backtrace: 0x78a43d7 0x78a422c 0x783c1b6 0x783acd0 0x783abb0 0x783ae4b 0x783c586 0x4251f 0x1745e5 0x5b012b0 0x5b017fd 0x5b01503 0x5b013f9 0x5b013c0 0x5b01131 0x5b00f21 0x5b00be2 0x5b00a4b 0x5b00978 0x5b000c4 0x5affd82 0x5affcc5 0x4251f 0x969fb 0x42475 0x287f2 0x2871a 0x39e95 0x4dc4091 0x4e3bc7b 0x4dc87a4 0x4fdebfc 0x4dba57e 0x4e1cff7 0x4e1cd94 0x7845d61 0x784731a 0x7848352 0x7847734 0x77aa304 0x77a9746 0x4b74c4e 0x29d8f 0x29e3f 0x4ad2fc4
 0# pthread_kill in /lib/x86_64-linux-gnu/libc.so.6
 1# raise in /lib/x86_64-linux-gnu/libc.so.6
 2# abort in /lib/x86_64-linux-gnu/libc.so.6
 3# 0x00007FD82A38D71B in /lib/x86_64-linux-gnu/libc.so.6
 4# 0x00007FD82A39EE96 in /lib/x86_64-linux-gnu/libc.so.6
 5# boost::intrusive_ptr<crimson::osd::ObjectContext>::operator->() const in /home/xuxuehan/src/ceph/clang-build/bin/crimson-osd
 6# crimson::osd::OpsExecuter::get_object_info() const in /home/xuxuehan/src/ceph/clang-build/bin/crimson-osd
 7# crimson::osd::OpsExecuter::get_target() const in /home/xuxuehan/src/ceph/clang-build/bin/crimson-osd
 8# crimson::osd::OpsExecuter::apply_stats() in /home/xuxuehan/src/ceph/clang-build/bin/crimson-osd
 9# crimson::interruptible::interruptible_future_detail<crimson::osd::IOInterruptCondition, seastar::future<std::tuple<crimson::interruptible::interruptible_future_detail<crimson::osd::IOInterruptCondition, seastar::future<void> >, crimson::interruptible::interruptible_future_detail<crimson::osd::IOInterruptCondition, seastar::future<void> > > > > crimson::osd::OpsExecuter::flush_changes_n_do_ops_effects<crimson::osd::PG::submit_executer(crimson::osd::OpsExecuter&&, std::vector<OSDOp, std::allocator<OSDOp> > const&)::$_0>(std::vector<OSDOp, std::allocator<OSDOp> > const&, SnapMapper&, OSDriver&, crimson::osd::PG::submit_executer(crimson::osd::OpsExecuter&&, std::vector<OSDOp, std::allocator<OSDOp> > const&)::$_0) && [clone .resume] at pg.cc:?
10# std::__n4861::coroutine_handle<crimson::internal::promise_base<crimson::interruptible::interruptor<crimson::osd::IOInterruptCondition>, void, std::tuple<crimson::interruptible::interruptible_future_detail<crimson::osd::IOInterruptCondition, seastar::future<void> >, crimson::interruptible::interruptible_future_detail<crimson::osd::IOInterruptCondition, seastar::future<void> > > > >::resume() const in /home/xuxuehan/src/ceph/clang-build/bin/crimson-osd
11# crimson::internal::promise_base<crimson::interruptible::interruptor<crimson::osd::IOInterruptCondition>, void, std::tuple<crimson::interruptible::interruptible_future_detail<crimson::osd::IOInterruptCondition, seastar::future<void> >, crimson::interruptible::interruptible_future_detail<crimson::osd::IOInterruptCondition, seastar::future<void> > > >::run_and_dispose() in /home/xuxuehan/src/ceph/clang-build/bin/crimson-osd
12# seastar::reactor::run_tasks(seastar::reactor::task_queue&) at :?
```

I've just added the commit back, please take a look, again. Thanks.

Reply:
c5a05faed82129daeb283cfdfda3e4d7e42c1e87 moved the apply_stats call there in order to allow for read ops to share stats as well.
By moving the call up we'll apply the stats before `flush_clone_metadata`, I think this should be ok as we adjust the clone's delta_stats in OpsExecuter ctor.